#### Consider the following problem:
In AB Company, there is a salary distribution table based on Year of Experience as given in
“salary_dataset.csv”. 
“The scenario is you are a HR officer, you got a candidate with 5 years of experience. Then what is the best salary you should offer to him?”

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('https://raw.githubusercontent.com/rahul96rajan/sample_datasets/master/salary_data.csv')
data.head()

YearsExperience  Salary
0              1.1   39343
1              1.3   46205
2              1.5   37731
3              2.0   43525
4              2.2   39891

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 2 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   YearsExperience  30 non-null     float64
 1   Salary           30 non-null     int64  
dtypes: float64(1), int64(1)
memory usage: 608.0 bytes


In [4]:
data.describe()

YearsExperience         Salary
count        30.000000      30.000000
mean          5.313333   76003.000000
std           2.837888   27414.429785
min           1.100000   37731.000000
25%           3.200000   56720.750000
50%           4.700000   65237.000000
75%           7.700000  100544.750000
max          10.500000  122391.000000

In [5]:
X = data['YearsExperience']
y = data['Salary']

In [6]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=1)

In [7]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import LeaveOneOut

loo = LeaveOneOut()

dtr = DecisionTreeRegressor()

param_grid = [{'max_leaf_nodes': list(range(2, 100)),
               'min_samples_split': [2, 3, 4]}]

grid_search = GridSearchCV(dtr, param_grid, cv=loo, scoring='neg_mean_squared_error',
                           return_train_score=True)
grid_search.fit(X_train.values.reshape(-1,1), y_train)
grid_search.best_estimator_

DecisionTreeRegressor(max_leaf_nodes=6)

In [8]:
# Training using the best_estimators_
dtr = DecisionTreeRegressor(max_leaf_nodes=6, random_state=1)
dtr.fit(X_train.values.reshape(-1,1), y_train)

DecisionTreeRegressor(max_leaf_nodes=6, random_state=1)

In [9]:
from sklearn.metrics import r2_score

y_pred = dtr.predict(X_test.values.reshape(-1,1))
y_pred_train = dtr.predict(X_train.values.reshape(-1,1))

print("Accuracy score (Train): ", r2_score(y_train, y_pred_train))
print("Accuracy score (Test): ", r2_score(y_test, y_pred))

Accuracy score (Train):  0.9901131953629431
Accuracy score (Test):  0.8268621052780349


In [10]:
# Now to find out salary for a candiate with a 5 year experience.

print(dtr.predict([[5]]))

[66983.5]


## <ins>Answer</ins>: So a candidate with 5 years of experince should be paid 66983.5 units of money.